### ISSUE TITLE
fail2ban doesn't work with python 3.9

### ISSUE THREAD

**Issue by opoplawski** [fail2ban/fail2ban#2610]
https://bugzilla.redhat.com/show_bug.cgi?id=1793476

            Bug ID: 1793476
           Summary: fail2ban fails to build with Python 3.9: imports abc
                    from collections
           Product: Fedora
           Version: rawhide
            Status: NEW
         Component: fail2ban
          Assignee: orion@nwra.com
          Reporter: mhroncok@redhat.com
        QA Contact: extras-qa@fedoraproject.org
                CC: anon.amish@gmail.com, axel.thimm@atrpms.net,
                    cstratak@redhat.com, mhroncok@redhat.com,
                    orion@nwra.com, vonsch@gmail.com
            Blocks: 1785415 (PYTHON39)
  Target Milestone: ---
    Classification: Fedora

fail2ban fails to build with Python 3.9.0a2.
```
+ /usr/bin/python3 bin/fail2ban-testcases --verbosity=2 --no-network
Fail2ban 0.11.1 test suite. Python 3.9.0a2 (default, Dec 19 2019, 00:00:00)
[GCC 9.2.1 20190827 (Red Hat 9.2.1-1)]. Please wait...
Traceback (most recent call last):
  File "/builddir/build/BUILD/fail2ban-0.11.1/bin/fail2ban-testcases", line 61,
in <module>
    tests = gatherTests(regexps, opts)
  File "./fail2ban/tests/utils.py", line 373, in gatherTests
    from . import clientreadertestcase
  File "./fail2ban/tests/clientreadertestcase.py", line 34, in <module>
    from ..client.jailreader import JailReader, extractOptions,
splitWithOptions
  File "./fail2ban/client/jailreader.py", line 34, in <module>
    from .actionreader import ActionReader
  File "./fail2ban/client/actionreader.py", line 31, in <module>
    from ..server.action import CommandAction
  File "./fail2ban/server/action.py", line 33, in <module>
    from collections import MutableMapping
ImportError: cannot import name 'MutableMapping' from 'collections'
(/usr/lib64/python3.9/collections/__init__.py)
```
See https://docs.python.org/3.9/whatsnew/3.9.html#removed

"The abstract base classes in collections.abc no longer are exposed in the
regular collections module. This will help create a clearer distinction between
the concrete classes and the abstract base classes."

For the build logs, see:
https://copr-be.cloud.fedoraproject.org/results/@python/python3.9/fedora-rawhide-x86_64/01170483-fail2ban/

For all our attempts to build fail2ban with Python 3.9, see:
https://copr.fedorainfracloud.org/coprs/g/python/python3.9/package/fail2ban/

Testing and mass rebuild of packages is happening in copr. You can follow these
instructions to test locally in mock if your package builds with Python 3.9:
https://copr.fedorainfracloud.org/coprs/g/python/python3.9/

Python 3.9 will be included in Fedora 33. To make that update smoother, we're
building Fedora packages with early pre-releases of Python 3.9.

**Comment by sebres** [comment]
I suppose TravisCI does not support 3.9-dev target yet...
I can try to fix it, but unless it doesn't supported by CI it'd be not advisable to use that.

> Python 3.9 will be included in Fedora 33.

As default python version (or just as dev)?

**Comment by opoplawski** [comment]
As the default (only) version.

**Comment by opoplawski** [comment]
I'll also note that F32 has yet to be released, so it's the release after that...

**Comment by opoplawski** [comment]
Apparently this change has been dropped from python 3.9, but will be coming in 3.10.

**Comment by sebres** [comment]
So let's reopen for 3.10 when the time comes.

**Comment by ftpd** [comment]
@sebres I have the same problem wit 3.10. Is it the time to reopen? ;-)

**Comment by sebres** [comment]
>  Is it the time to reopen?

I don't think so. The newest version(s) is running under 3.10 (for instance see the CI reports on https://github.com/fail2ban/fail2ban/actions).
There was already another issue(s) targeting that (#2931, #3142), if you have the same - then it is already fixed (in 294ec73f629d0e29cece3a1eb5dd60b6fccea41f).
This was not yet released here, but the fix is trivial, so @fail2ban/maintainers could made a simple hot-fix if targeting python 3.10 distribution.
I'm unsure whether there is any distro which has packaged 3.10 by default (or fail2ban is build there for 3.10), is it?

**Comment by ftpd** [comment]
Void has 3.10 and I'm using it. The fail2ban version available in repos is 0.11.2 and it seems it doesn't include this fix, am I right?

**Comment by sebres** [comment]
Stock (upstream) 0.11.2 doesn't include it, right... but I don't know about this distro package (whether it may be hot-fixed or not). However if you see the error from #2931 (`cannot import name '*Mapping' from 'collections'`) it is obvious that it doesn't really.
Just I wonder how they build this package with 3.10 (build process normally imply running of test-suite and the tests would fail).

**Comment by ftpd** [comment]
I built it yesterday myself and the build passed. I'll do it once again and provide build log.

**Comment by sebres** [comment]
Then your build-script doesn't invoke `bin/fail2ban-testcases`, but you could do it by yourselves also for uninstalled fresh built folder:
`PYTHONPATH=. python3.10 bin/fail2ban-testcases` (replace `.` with prefix for the modules in build if it is not current folder).
Anyway the fix was made after 0.11.2 got released, so it'd definitely fail.
You can also checkout current 0.11 branch (containing this fix), since we're holding released branches (0.10/0.11) stable, and built it from source, but I don't know which patches/includes are additionally provided by the maintainers for Void.
But a comparison between upstream 0.11.2 and the sources you was building from would show the differences.

Alternatively simply apply 294ec73f629d0e29cece3a1eb5dd60b6fccea41f as patch for the 0.11.2 source before you build.

**Comment by opoplawski** [comment]
FWIW - Fedora 35 (just released) uses python 3.10 as the standard python version.  Of course, we are carrying the patch.
